src/hooks/useSafeState.ts

```diff
@@ -0,0 +1,28 @@
+import {
+  useState,
+  useCallback,
+  useRef,
+  useEffect,
+  SetStateAction,
+  Dispatch,
+} from "react";
+
+export default function useSafeState<T>(
```

Comment:
Safe state is to prevent us from setting state on an unmounted component, right? Or is there more to it? Also is this a big problem, and why hasn't React made updates to handle this better? Are we doing something weird?

Reply:
It can definitely be a big problem, especially if you're dealing with components that might be quickly mounting and unmounting, like in the case of a SPA.

I'm not exactly sure why React hasn't done something to specifically mitigate this problem. With hooks and functional components, React doesn't explicitly track the "mountedness" of a component, and since it's all just javascript, it can be really easy to have a request start, a component unmount, and then for your program to break trying to update state when the request finishes and tries to keep running code. So something like this is important in a general-use tool because you can't guarantee that it will be used a certain way downstream.